core/src/main/java/com/datastax/oss/driver/api/core/PagingIterable.java

```diff
@@ -0,0 +1,161 @@
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Function;
+
+/**
+ * An iterable of elements which are fetched synchronously by the driver, possibly in multiple
+ * requests.
+ *
+ * <p>It uses asynchronous calls internally, but blocks on the results in order to provide a
+ * synchronous API to its clients. If the query is paged, only the first page will be fetched
+ * initially, and iteration will trigger background fetches of the next pages when necessary.
+ *
+ * <p>Note that this object can only be iterated once: elements are "consumed" as they are read,
+ * subsequent calls to {@code iterator()} will return the same iterator instance.
+ *
+ * <p>Implementations of this type are <b>not thread-safe</b>. They can only be iterated by the
+ * thread that invoked {@code session.execute}.
+ *
+ * <p>This is a generalization of {@link ResultSet}, replacing rows by an arbitrary element type.
+ */
+public interface PagingIterable<ElementT> extends Iterable<ElementT> {
+
+  /** Metadata about the columns returned by the CQL request that was used to build this result. */
+  @NonNull
+  ColumnDefinitions getColumnDefinitions();
+
+  /**
+   * The execution information for the last query performed for this iterable.
+   *
+   * <p>This is a shortcut for:
+   *
+   * <pre>
+   * getExecutionInfos().get(getExecutionInfos().size() - 1)
+   * </pre>
+   *
+   * @see #getExecutionInfos()
+   */
+  @NonNull
+  default ExecutionInfo getExecutionInfo() {
+    List<ExecutionInfo> infos = getExecutionInfos();
+    return infos.get(infos.size() - 1);
+  }
+
+  /**
+   * The execution information for all the queries that have been performed so far to assemble this
+   * iterable.
+   *
+   * <p>This will have multiple elements if the query is paged, since the driver performs blocking
+   * background queries to fetch additional pages transparently as the result set is being iterated.
+   */
+  @NonNull
+  List<ExecutionInfo> getExecutionInfos();
+
+  /**
+   * Returns the next element, or {@code null} if the iterable is exhausted.
+   *
+   * <p>This is convenient for queries that are known to return exactly one row, for example count
+   * queries.
+   */
+  @Nullable
+  default ElementT one() {
+    Iterator<ElementT> iterator = iterator();
+    return iterator.hasNext() ? iterator.next() : null;
+  }
+
+  /**
+   * Returns all the remaining elements as a list; <b>not recommended for queries that return a
+   * large number of elements</b>.
+   *
+   * <p>Contrary to {@link #iterator()} or successive calls to {@link #one()}, this method forces
+   * fetching the <b>full contents</b> at once; in particular, this means that a large number of
+   * background queries might have to be run, and that all the data will be held in memory locally.
+   * Therefore it is crucial to only call this method for queries that are known to return a
+   * reasonable number of results.
+   */
```

Comment:
Would it make sense to to checks in the actual implementation to prevent blowing out the heap... or is this just an instance of user better know what they are doing?

Reply:
We can't predict in advance the number of rows, or even the size each row will take in memory, so it's more the latter IMHO.
This method is already present in 3.x so users should be familiar with the pitfalls.